File: Java/problem454.java
class Solution {
    public int fourSumCount(int[] A, int[] B, int[] C, int[] D) {
        int count = 0, num;
        Map<Integer, Integer> map = new HashMap<>();
        for (int a : A) {
            for (int b : B) {
                num = 0 - a - b;
                if (map.containsKey(num)) {
                    map.put(num, map.get(num) + 1);
                }
                else {
                    map.put(num, 1);
                }
            }
        }
        for (int c : C) {
            for (int d : D) {
                num = c + d;
                if (map.containsKey(num)) {
                    count += map.get(num);
                }
            }
        }
        return count;
    }
}

File: python/problem454.py
class Solution:

    def fourSumCount(self, A: List[int], B: List[int], C: List[int], D: List[int]) -> int:
        residual2num = {}
        count = 0
        for a in A:
            for b in B:
                num = 0 - a - b
                if num in residual2num:     #整合A和B各个数的组合，以0-a-b作为key，出现次数作为value
                    residual2num[num] += 1
                else:
                    residual2num[num] = 1
        for c in C:
            for d in D:
                num = c + d
                if num in residual2num:     #只要c+d出现在字典中，说明出现一个符合的四元组
                    count += residual2num[num]
        return count
